File: src/ker_cpp/ker_benchmark.cpp
// Filename: src/ker_cpp/ker_benchmark.cpp

#include "ker_kernel.cpp"
#include <chrono>
#include <random>
#include <vector>

int main() {
    constexpr std::size_t iterations = 1000;
    std::array<double, NUM_PLANES> r{};
    std::array<double, NUM_PLANES> w{};

    std::mt19937_64 rng(123456);
    std::uniform_real_distribution<double> dist(0.0, 1.0);

    for (std::size_t i = 0; i < NUM_PLANES; ++i) {
        w[i] = dist(rng);
    }

    double vtPrev = 0.0;

    auto t0 = std::chrono::high_resolution_clock::now();
    for (std::size_t it = 0; it < iterations; ++it) {
        for (std::size_t i = 0; i < NUM_PLANES; ++i) {
            r[i] = dist(rng);
        }
        KerResult res = computeKer(r, w, vtPrev);
        vtPrev = res.vt;
    }
    auto t1 = std::chrono::high_resolution_clock::now();

    auto dtNs = std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();
    // On Raspberry Pi, energy measurement is done externally (e.g. via power meter),
    // this program produces consistent compute patterns for correlation.
    return 0;
}

File: src/ker_java/KerBenchmark.java
// Filename: src/ker_java/KerBenchmark.java

package ker_java;

import java.util.Random;

public final class KerBenchmark {

    public static void main(String[] args) {
        final int iterations = 1000;
        final double[] r = new double[KerKernel.NUM_PLANES];
        final double[] w = new double[KerKernel.NUM_PLANES];
        final Random rng = new Random(123456L);

        for (int i = 0; i < KerKernel.NUM_PLANES; i++) {
            w[i] = rng.nextDouble();
        }

        double vtPrev = 0.0;

        long t0 = System.nanoTime();
        for (int it = 0; it < iterations; it++) {
            for (int i = 0; i < KerKernel.NUM_PLANES; i++) {
                r[i] = rng.nextDouble();
            }
            KerKernel.KerResult res = KerKernel.computeKer(r, w, vtPrev);
            vtPrev = res.vt;
        }
        long t1 = System.nanoTime();

        long dtNs = t1 - t0;
        // Similar to C++, external measurement captures energy;
        // dtNs allows correlating compute time with energy.
    }
}
